src/psyclone/psyGen.py

```diff
@@ -2988,6 +3025,198 @@ def ast(self):
         self._fp2_ast = my_parser(reader)
         return self._fp2_ast
 
+    @staticmethod
+    def _new_name(original, tag, suffix):
+        '''
+        Construct a new name given the original, a tag and a suffix (which
+        may or may not terminate the original name). If suffix is present
+        in the original name then the `tag` is inserted before it.
+
+        :param str original: The original name
+        :param str tag: Tag to insert into new name
+        :param str suffix: Suffix with which to end new name.
+        :returns: New name made of original + tag + suffix
+        :rtype: str
+        '''
+        if original.endswith(suffix):
+            return original[:-len(suffix)] + tag + suffix
+        return original + tag + suffix
+
+    def update(self):
+        '''
+        Writes the (transformed) AST of this kernel to file and resets the
+        'modified' flag to False. By default, the kernel is re-named
+        so as to be unique within the kernel output directory stored
+        within the configuration object. If config.kernel_naming is
+        "single" then no re-naming and output is performed if there is
+        already a transformed copy of the kernel in the output dir.
+
+        :raises GenerationError: if config.kernel_naming == "single" and a \
+                                 different, transformed version of this \
+                                 kernel is already in the output directory.
+        '''
+        import os
+        from psyclone.line_length import FortLineLength
+
+        # If this kernel has not been transformed we do nothing
+        if not self.modified:
+            return
+
+        # Remove any "_mod" if the file follows the PSyclone naming convention
+        orig_mod_name = self.module_name[:]
+        if orig_mod_name.endswith("_mod"):
+            old_base_name = orig_mod_name[:-4]
+        else:
+            old_base_name = orig_mod_name[:]
+
+        # We could create a hash of a string built from the name of the
+        # Algorithm (module), the name/position of the Invoke and the
+        # index of this kernel within that Invoke. However, that creates
+        # a very long name so we simply ensure that kernel names are unique
+        # within the user-supplied kernel-output directory.
+        name_idx = -1
+        fdesc = None
+        while not fdesc:
+            name_idx += 1
```

Comment:
Presumably it does not make sense to use the namespace manager here?

Reply:
No, I don't think it does since we have to use the filesystem to manage our names.